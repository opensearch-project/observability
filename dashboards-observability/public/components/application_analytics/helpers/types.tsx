/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export interface AvailabilityType {
  name: string;
  color: string;
  mainVisId: string;
}
