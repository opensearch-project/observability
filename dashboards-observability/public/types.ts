/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { DashboardStart } from '../../../src/plugins/dashboard/public';
import { NavigationPublicPluginStart } from '../../../src/plugins/navigation/public';

export interface AppPluginStartDependencies {
  navigation: NavigationPublicPluginStart;
  dashboard: DashboardStart;
}

export interface ObservabilitySetup {}

export interface ObservabilityStart {}
