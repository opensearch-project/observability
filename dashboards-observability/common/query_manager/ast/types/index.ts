/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export {
  ExpressionChunk,
  SpanChunk,
  StatsAggregationChunk,
  StatsAggregationFunctionChunk,
  GroupByChunk,
  GroupField,
  StatsChunk,
  SpanExpressionChunk,
  AggregationConfigurations,
  PreviouslyParsedStaleStats
} from './stats';
