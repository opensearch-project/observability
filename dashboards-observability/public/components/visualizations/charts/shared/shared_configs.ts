/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { forIn } from 'lodash';

export const getPlotlySharedConfigs = () => {
  return {
    layout: {
      showlegend: true,
      margin: {
        l: 60,
        r: 30,
        b: 30,
        t: 50,
        pad: 0,
      },
      height: 1180,
      legend: {
        orientation: 'v',
        traceorder: 'normal',
      },
    },
    config: {
      displaylogo: false,
      responsive: true,
    },
  };
};

enum VIS_CATEGORY {
  BASICS = 'Visualizations',
}

export const getPlotlyCategory = (type = 'enum') => {
  switch (type) {
    case 'enum':
      return VIS_CATEGORY;
    case 'list':
      return Object.values(VIS_CATEGORY);
    case 'keys':
      return Object.keys(VIS_CATEGORY);
    default:
      return VIS_CATEGORY;
  }
};
