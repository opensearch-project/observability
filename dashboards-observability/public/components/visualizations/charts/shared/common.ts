/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { isEmpty, forEach } from 'lodash';
import { CUSTOM_LABEL } from '../../../../../common/constants/explorer';
import { ConfigList, DimensionSpan } from '../../../../../common/types/explorer';
import { removeBacktick } from '../../../../../common/utils';

export const getCompleteTimespanKey = (span: DimensionSpan) => {
  if (isEmpty(span) || isEmpty(span.time_field) || isEmpty(span.interval) || isEmpty(span.unit))
    return '';
  const spanName = `span(${span.time_field[0]?.name},${span.interval}${span.unit[0]?.value})`;
  return { name: spanName, label: spanName };
};

/**
 * Transform to traces that can be consumed by plotly.
 * @param intermediateVisData preprocessed json data that has dimensions to single aggregation mapping.
 * @param param1 required visualization configurations.
 * @returns traces.
 */
export const transformPreprocessedDataToTraces = (
  intermediateVisData: Array<any>,
  { breakdowns, isVertical = true }: Partial<ConfigList>
) => {
  const traceMap = new Map<string, any>();
  const hasBreakdown = !isEmpty(breakdowns);
  forEach(intermediateVisData, (entry) => {
    const traceKey = hasBreakdown ? [entry.breakdown, entry.aggName].join(',') : entry.aggName;
    if (isEmpty(traceMap.get(traceKey))) {
      traceMap.set(traceKey, {
        x: isVertical ? [entry.x] : [entry.value],
        y: isVertical ? [entry.value] : [entry.x],
        name: hasBreakdown ? [entry.breakdown, entry.aggName].join(',') : `${traceKey}`,
      });
    } else {
      const curTrace = traceMap.get(traceKey);
      const xaxisValue = isVertical ? entry.x : entry.value;
      const yaxisValue = isVertical ? entry.value : entry.x;
      curTrace!.x.push(xaxisValue);
      curTrace!.y.push(yaxisValue);
    }
  });
  return [...traceMap.values()];
};

/**
 * preprocess json data to
 * 1. concatenate dimensions to generate one dimension
 * 2. concatenate breakdowns (if there's any) generate one breakdown
 * 3. map dimension/breakdown to aggregations
 * @param visJson raw json data from data fetching
 * @param param1 required visualization configurations.
 * @returns intermediate visualization json data
 */
export const preprocessJsonData = (
  visJson: Array<any>,
  { dimensions, series, breakdowns, span }: Partial<ConfigList>
) => {
  const seriesFlattenedEntries = [];
  forEach(visJson, (entry) => {
    const backtickRemovedEntry = {};
    // remove backtick, so data in jsonData can be accessed through using field name
    forEach(entry, (value, key) => {
      backtickRemovedEntry[removeBacktick(key)] = value;
    });
    forEach(series, (sr) => {
      let tabularVizData = {};
      const serieKey = sr[CUSTOM_LABEL] ? sr[CUSTOM_LABEL] : `${sr.aggregation}(${sr.name})`;
      if (!isEmpty(serieKey)) {
        const concatedXaxisLabel = [
          ...(!isEmpty(span) ? [getCompleteTimespanKey(span)] : []),
          ...dimensions,
        ]
          .map((dimension) => {
            return backtickRemovedEntry[removeBacktick(dimension.name)] ?? '';
          })
          .join(',');
        const concatedBreakdownLabel = breakdowns
          ? breakdowns
              .map((breakdown) => backtickRemovedEntry[removeBacktick(breakdown.name)])
              .join(',')
          : '';
        tabularVizData = {
          value: backtickRemovedEntry[serieKey],
          x: concatedXaxisLabel,
          breakdown: concatedBreakdownLabel,
          aggName: serieKey,
        };
      } else {
        tabularVizData = {
          value: 0,
          x: '',
          breakdown: '',
        };
      }
      seriesFlattenedEntries.push(tabularVizData);
    });
  });
  return seriesFlattenedEntries;
};
