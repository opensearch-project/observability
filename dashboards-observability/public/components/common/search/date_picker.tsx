/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { EuiSuperDatePicker } from '@elastic/eui';
import { IDatePickerProps } from './search';
import { uiSettingsService } from '../../../../common/utils';

export function DatePicker(props: IDatePickerProps) {
  const { startTime, endTime, handleTimePickerChange, handleTimeRangePickerRefresh } = props;

  const handleTimeChange = (e: any) => handleTimePickerChange([e.start, e.end]);

  return (
    <EuiSuperDatePicker
      data-test-subj="pplSearchDatePicker"
      start={startTime}
      end={endTime}
      dateFormat={uiSettingsService.get('dateFormat')}
      onTimeChange={handleTimeChange}
      onRefresh={handleTimeRangePickerRefresh}
      className="osdQueryBar__datePicker"
    />
  );
}
