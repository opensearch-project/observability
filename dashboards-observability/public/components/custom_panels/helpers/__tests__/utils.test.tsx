/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import moment from 'moment';
import { DurationRange } from '@elastic/eui/src/components/date_picker/types';

import {
  isNameValid,
  convertDateTime,
  mergeLayoutAndVisualizations,
  onTimeChange,
  isDateValid,
  isPPLFilterValid,
  displayVisualization,
} from '../utils';
import {
  sampleLayout,
  sampleMergedVisualizations,
  samplePanelVisualizations,
  samplePPLEmptyResponse,
  samplePPLResponse,
  sampleSavedVisualization,
  sampleSavedVisualizationForHorizontalBar,
  sampleSavedVisualizationForLine,
  sampleSavedVisualizationForPie,
  sampleSavedVisualizationForTreeMap,
} from '../../../../../test/panels_constants';
import { PPL_DATE_FORMAT } from '../../../../../common/constants/shared';
import React from 'react';

describe('Utils helper functions', () => {
  configure({ adapter: new Adapter() });

  it('validates isNameValid function', () => {
    expect(isNameValid('Lorem ipsum dolor sit amet, consectetur adipiscing elit,')).toBe(false);
    expect(isNameValid('Lorem ipsum dolor sit amet, consectetur adipiscin')).toBe(true);
  });

  it('validates convertDateTime function', () => {
    expect(convertDateTime('2022-01-30T18:44:40.577Z')).toBe(
      moment('2022-01-30T18:44:40.577Z').format(PPL_DATE_FORMAT)
    );
    expect(convertDateTime('2022-02-25T19:18:33.075Z', true)).toBe(
      moment('2022-02-25T19:18:33.075Z').format(PPL_DATE_FORMAT)
    );
  });

  it('validates mergeLayoutAndVisualizations function', () => {
    const setState = jest.fn();
    mergeLayoutAndVisualizations(sampleLayout, samplePanelVisualizations, setState);
    expect(setState).toHaveBeenCalledWith(sampleMergedVisualizations);
  });

  it('validates onTimeChange function', () => {
    const setRecentlyUsedRanges = jest.fn((x) => x);
    const setStart = jest.fn();
    const setEnd = jest.fn();
    const recentlyUsedRanges: DurationRange[] = [];
    onTimeChange(
      '2022-01-30T18:44:40.577Z',
      '2022-02-25T19:18:33.075Z',
      recentlyUsedRanges,
      setRecentlyUsedRanges,
      setStart,
      setEnd
    );
    expect(setRecentlyUsedRanges).toHaveBeenCalledWith([
      { start: '2022-01-30T18:44:40.577Z', end: '2022-02-25T19:18:33.075Z' },
    ]);
    expect(setStart).toHaveBeenCalledWith('2022-01-30T18:44:40.577Z');
    expect(setEnd).toHaveBeenCalledWith('2022-02-25T19:18:33.075Z');
  });

  it('validates isDateValid function', () => {
    const setToast = jest.fn();
    expect(
      isDateValid(
        convertDateTime('2022-01-30T18:44:40.577Z'),
        convertDateTime('2022-02-25T19:18:33.075Z', false),
        setToast
      )
    ).toBe(true);
    expect(
      isDateValid(
        convertDateTime('2022-01-30T18:44:40.577Z'),
        convertDateTime('2022-01-30T18:44:40.577Z', false),
        setToast
      )
    ).toBe(true);
    expect(
      isDateValid(
        convertDateTime('2022-02-25T19:18:33.075Z'),
        convertDateTime('2022-01-30T18:44:40.577Z', false),
        setToast
      )
    ).toBe(false);
  });

  it('validates isPPLFilterValid function', () => {
    const setToast = jest.fn();
    expect(isPPLFilterValid(sampleSavedVisualization.visualization.query, setToast)).toBe(false);
    expect(isPPLFilterValid("where Carrier = 'OpenSearch-Air'", setToast)).toBe(true);
  });

  it('renders displayVisualization function', () => {
    const wrapper1 = mount(
      <div>
        {displayVisualization(sampleSavedVisualization.visualization, samplePPLResponse, 'bar')}
      </div>
    );
    expect(wrapper1).toMatchSnapshot();

    const wrapper2 = mount(
      <div>{displayVisualization(sampleSavedVisualizationForLine, samplePPLResponse, 'line')}</div>
    );
    expect(wrapper2).toMatchSnapshot();

    const wrapper4 = mount(
      <div>
        {displayVisualization(
          sampleSavedVisualizationForHorizontalBar,
          samplePPLResponse,
          'horizontal_bar'
        )}
      </div>
    );
    expect(wrapper4).toMatchSnapshot();

    const wrapper6 = mount(
      <div>{displayVisualization({}, samplePPLEmptyResponse, 'horizontal_bar')}</div>
    );
    expect(wrapper6).toMatchSnapshot();
  });
});
