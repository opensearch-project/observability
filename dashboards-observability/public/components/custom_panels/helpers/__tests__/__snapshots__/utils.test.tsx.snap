// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils helper functions renders displayVisualization function 1`] = `
<div>
  <Bar
    layoutConfig={
      Object {
        "xaxis": Object {
          "fixedrange": false,
          "showgrid": false,
        },
        "yaxis": Object {
          "fixedrange": false,
          "showgrid": false,
        },
      }
    }
    visualizations={
      Object {
        "data": Object {
          "Carrier": Array [
            "BeatsWest",
            "Logstash Airways",
            "OpenSearch Dashboards Airlines",
            "OpenSearch-Air",
          ],
          "avg(FlightDelayMin)": Array [
            53.65384615384615,
            45.36144578313253,
            63.1578947368421,
            46.81318681318681,
          ],
        },
        "jsonData": Array [
          Object {
            "Carrier": "BeatsWest",
            "avg(FlightDelayMin)": 53.65384615384615,
          },
          Object {
            "Carrier": "Logstash Airways",
            "avg(FlightDelayMin)": 45.36144578313253,
          },
          Object {
            "Carrier": "OpenSearch Dashboards Airlines",
            "avg(FlightDelayMin)": 63.1578947368421,
          },
          Object {
            "Carrier": "OpenSearch-Air",
            "avg(FlightDelayMin)": 46.81318681318681,
          },
        ],
        "metadata": Object {
          "fields": Array [
            Object {
              "name": "avg(FlightDelayMin)",
              "type": "double",
            },
            Object {
              "name": "Carrier",
              "type": "keyword",
            },
          ],
        },
        "size": 4,
        "status": 200,
      }
    }
  >
    <Plt
      config={Object {}}
      data={
        Array [
          Object {
            "marker": Object {
              "color": Array [
                "#3CA1C7",
                "#8C55A3",
                "#DB748A",
                "#F2BE4B",
              ],
            },
            "name": "avg(FlightDelayMin)",
            "type": "bar",
            "x": Array [
              "BeatsWest",
              "Logstash Airways",
              "OpenSearch Dashboards Airlines",
              "OpenSearch-Air",
            ],
            "y": Array [
              53.65384615384615,
              45.36144578313253,
              63.1578947368421,
              46.81318681318681,
            ],
          },
        ]
      }
      layout={
        Object {
          "colorway": Array [
            "#3CA1C7",
            "#8C55A3",
            "#DB748A",
            "#F2BE4B",
            "#68CCC2",
            "#2A7866",
            "#843769",
            "#374FB8",
            "#BD6F26",
            "#4C636F",
          ],
          "xaxis": Object {
            "automargin": true,
            "fixedrange": false,
            "showgrid": false,
          },
          "yaxis": Object {
            "automargin": true,
            "fixedrange": false,
            "showgrid": false,
          },
        }
      }
    >
      <PlotlyComponent
        config={
          Object {
            "displayModeBar": false,
          }
        }
        data={
          Array [
            Object {
              "marker": Object {
                "color": Array [
                  "#3CA1C7",
                  "#8C55A3",
                  "#DB748A",
                  "#F2BE4B",
                ],
              },
              "name": "avg(FlightDelayMin)",
              "type": "bar",
              "x": Array [
                "BeatsWest",
                "Logstash Airways",
                "OpenSearch Dashboards Airlines",
                "OpenSearch-Air",
              ],
              "y": Array [
                53.65384615384615,
                45.36144578313253,
                63.1578947368421,
                46.81318681318681,
              ],
            },
          ]
        }
        debug={false}
        layout={
          Object {
            "autosize": true,
            "barmode": "stack",
            "colorway": Array [
              "#3CA1C7",
              "#8C55A3",
              "#DB748A",
              "#F2BE4B",
              "#68CCC2",
              "#2A7866",
              "#843769",
              "#374FB8",
              "#BD6F26",
              "#4C636F",
            ],
            "hovermode": "closest",
            "legend": Object {
              "orientation": "h",
              "traceorder": "normal",
            },
            "margin": Object {
              "b": 30,
              "l": 30,
              "pad": 4,
              "r": 5,
              "t": 5,
            },
            "showlegend": false,
            "xaxis": Object {
              "automargin": true,
              "fixedrange": false,
              "showgrid": false,
            },
            "yaxis": Object {
              "automargin": true,
              "fixedrange": false,
              "showgrid": false,
            },
          }
        }
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
        useResizeHandler={true}
      >
        <div
          style={
            Object {
              "height": "100%",
              "width": "100%",
            }
          }
        />
      </PlotlyComponent>
    </Plt>
  </Bar>
</div>
`;

exports[`Utils helper functions renders displayVisualization function 2`] = `
<div>
  <Line
    layoutConfig={
      Object {
        "xaxis": Object {
          "fixedrange": true,
          "showgrid": false,
        },
        "yaxis": Object {
          "fixedrange": true,
          "showgrid": false,
        },
      }
    }
    visualizations={
      Object {
        "data": Object {
          "Carrier": Array [
            "BeatsWest",
            "Logstash Airways",
            "OpenSearch Dashboards Airlines",
            "OpenSearch-Air",
          ],
          "avg(FlightDelayMin)": Array [
            53.65384615384615,
            45.36144578313253,
            63.1578947368421,
            46.81318681318681,
          ],
        },
        "jsonData": Array [
          Object {
            "Carrier": "BeatsWest",
            "avg(FlightDelayMin)": 53.65384615384615,
          },
          Object {
            "Carrier": "Logstash Airways",
            "avg(FlightDelayMin)": 45.36144578313253,
          },
          Object {
            "Carrier": "OpenSearch Dashboards Airlines",
            "avg(FlightDelayMin)": 63.1578947368421,
          },
          Object {
            "Carrier": "OpenSearch-Air",
            "avg(FlightDelayMin)": 46.81318681318681,
          },
        ],
        "metadata": Object {
          "fields": Array [
            Object {
              "name": "avg(FlightDelayMin)",
              "type": "double",
            },
            Object {
              "name": "Carrier",
              "type": "keyword",
            },
          ],
        },
        "size": 4,
        "status": 200,
      }
    }
  >
    <Plt
      config={Object {}}
      data={
        Array [
          Object {
            "name": "avg(FlightDelayMin)",
            "type": "line",
            "x": Array [
              "BeatsWest",
              "Logstash Airways",
              "OpenSearch Dashboards Airlines",
              "OpenSearch-Air",
            ],
            "y": Array [
              53.65384615384615,
              45.36144578313253,
              63.1578947368421,
              46.81318681318681,
            ],
          },
        ]
      }
      layout={
        Object {
          "barmode": "line",
          "colorway": Array [
            "#3CA1C7",
            "#8C55A3",
            "#DB748A",
            "#F2BE4B",
            "#68CCC2",
            "#2A7866",
            "#843769",
            "#374FB8",
            "#BD6F26",
            "#4C636F",
          ],
          "paper_bgcolor": "rgba(0, 0, 0, 0)",
          "plot_bgcolor": "rgba(0, 0, 0, 0)",
          "xaxis": Object {
            "automargin": true,
            "fixedrange": true,
            "showgrid": false,
          },
          "yaxis": Object {
            "automargin": true,
            "fixedrange": true,
            "showgrid": false,
          },
        }
      }
    >
      <PlotlyComponent
        config={
          Object {
            "displayModeBar": false,
          }
        }
        data={
          Array [
            Object {
              "name": "avg(FlightDelayMin)",
              "type": "line",
              "x": Array [
                "BeatsWest",
                "Logstash Airways",
                "OpenSearch Dashboards Airlines",
                "OpenSearch-Air",
              ],
              "y": Array [
                53.65384615384615,
                45.36144578313253,
                63.1578947368421,
                46.81318681318681,
              ],
            },
          ]
        }
        debug={false}
        layout={
          Object {
            "autosize": true,
            "barmode": "line",
            "colorway": Array [
              "#3CA1C7",
              "#8C55A3",
              "#DB748A",
              "#F2BE4B",
              "#68CCC2",
              "#2A7866",
              "#843769",
              "#374FB8",
              "#BD6F26",
              "#4C636F",
            ],
            "hovermode": "closest",
            "legend": Object {
              "orientation": "h",
              "traceorder": "normal",
            },
            "margin": Object {
              "b": 30,
              "l": 30,
              "pad": 4,
              "r": 5,
              "t": 5,
            },
            "paper_bgcolor": "rgba(0, 0, 0, 0)",
            "plot_bgcolor": "rgba(0, 0, 0, 0)",
            "showlegend": false,
            "xaxis": Object {
              "automargin": true,
              "fixedrange": true,
              "showgrid": false,
            },
            "yaxis": Object {
              "automargin": true,
              "fixedrange": true,
              "showgrid": false,
            },
          }
        }
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
        useResizeHandler={true}
      >
        <div
          style={
            Object {
              "height": "100%",
              "width": "100%",
            }
          }
        />
      </PlotlyComponent>
    </Plt>
  </Line>
</div>
`;

exports[`Utils helper functions renders displayVisualization function 3`] = `
<div>
  <HorizontalBar
    layoutConfig={
      Object {
        "xaxis": Object {
          "fixedrange": false,
          "showgrid": false,
        },
        "yaxis": Object {
          "fixedrange": false,
          "showgrid": false,
        },
      }
    }
    visualizations={
      Object {
        "data": Object {
          "Carrier": Array [
            "BeatsWest",
            "Logstash Airways",
            "OpenSearch Dashboards Airlines",
            "OpenSearch-Air",
          ],
          "avg(FlightDelayMin)": Array [
            53.65384615384615,
            45.36144578313253,
            63.1578947368421,
            46.81318681318681,
          ],
        },
        "jsonData": Array [
          Object {
            "Carrier": "BeatsWest",
            "avg(FlightDelayMin)": 53.65384615384615,
          },
          Object {
            "Carrier": "Logstash Airways",
            "avg(FlightDelayMin)": 45.36144578313253,
          },
          Object {
            "Carrier": "OpenSearch Dashboards Airlines",
            "avg(FlightDelayMin)": 63.1578947368421,
          },
          Object {
            "Carrier": "OpenSearch-Air",
            "avg(FlightDelayMin)": 46.81318681318681,
          },
        ],
        "metadata": Object {
          "fields": Array [
            Object {
              "name": "avg(FlightDelayMin)",
              "type": "double",
            },
            Object {
              "name": "Carrier",
              "type": "keyword",
            },
          ],
        },
        "size": 4,
        "status": 200,
      }
    }
  >
    <Bar
      barConfig={
        Object {
          "orientation": "h",
        }
      }
      layoutConfig={
        Object {
          "xaxis": Object {
            "fixedrange": false,
            "showgrid": false,
          },
          "yaxis": Object {
            "automargin": true,
            "fixedrange": false,
            "showgrid": false,
          },
        }
      }
      visualizations={
        Object {
          "data": Object {
            "Carrier": Array [
              "BeatsWest",
              "Logstash Airways",
              "OpenSearch Dashboards Airlines",
              "OpenSearch-Air",
            ],
            "avg(FlightDelayMin)": Array [
              53.65384615384615,
              45.36144578313253,
              63.1578947368421,
              46.81318681318681,
            ],
          },
          "jsonData": Array [
            Object {
              "Carrier": "BeatsWest",
              "avg(FlightDelayMin)": 53.65384615384615,
            },
            Object {
              "Carrier": "Logstash Airways",
              "avg(FlightDelayMin)": 45.36144578313253,
            },
            Object {
              "Carrier": "OpenSearch Dashboards Airlines",
              "avg(FlightDelayMin)": 63.1578947368421,
            },
            Object {
              "Carrier": "OpenSearch-Air",
              "avg(FlightDelayMin)": 46.81318681318681,
            },
          ],
          "metadata": Object {
            "fields": Array [
              Object {
                "name": "avg(FlightDelayMin)",
                "type": "double",
              },
              Object {
                "name": "Carrier",
                "type": "keyword",
              },
            ],
          },
          "size": 4,
          "status": 200,
        }
      }
    >
      <Plt
        config={
          Object {
            "orientation": "h",
          }
        }
        data={
          Array [
            Object {
              "marker": Object {
                "color": Array [
                  "#3CA1C7",
                  "#8C55A3",
                  "#DB748A",
                  "#F2BE4B",
                ],
              },
              "name": "avg(FlightDelayMin)",
              "orientation": "h",
              "type": "bar",
              "x": Array [
                53.65384615384615,
                45.36144578313253,
                63.1578947368421,
                46.81318681318681,
              ],
              "y": Array [
                "BeatsWest",
                "Logstash Airways",
                "OpenSearch Dashboards Airlines",
                "OpenSearch-Air",
              ],
            },
          ]
        }
        layout={
          Object {
            "colorway": Array [
              "#3CA1C7",
              "#8C55A3",
              "#DB748A",
              "#F2BE4B",
              "#68CCC2",
              "#2A7866",
              "#843769",
              "#374FB8",
              "#BD6F26",
              "#4C636F",
            ],
            "xaxis": Object {
              "automargin": true,
              "fixedrange": false,
              "showgrid": false,
            },
            "yaxis": Object {
              "automargin": true,
              "fixedrange": false,
              "showgrid": false,
            },
          }
        }
      >
        <PlotlyComponent
          config={
            Object {
              "displayModeBar": false,
              "orientation": "h",
            }
          }
          data={
            Array [
              Object {
                "marker": Object {
                  "color": Array [
                    "#3CA1C7",
                    "#8C55A3",
                    "#DB748A",
                    "#F2BE4B",
                  ],
                },
                "name": "avg(FlightDelayMin)",
                "orientation": "h",
                "type": "bar",
                "x": Array [
                  53.65384615384615,
                  45.36144578313253,
                  63.1578947368421,
                  46.81318681318681,
                ],
                "y": Array [
                  "BeatsWest",
                  "Logstash Airways",
                  "OpenSearch Dashboards Airlines",
                  "OpenSearch-Air",
                ],
              },
            ]
          }
          debug={false}
          layout={
            Object {
              "autosize": true,
              "barmode": "stack",
              "colorway": Array [
                "#3CA1C7",
                "#8C55A3",
                "#DB748A",
                "#F2BE4B",
                "#68CCC2",
                "#2A7866",
                "#843769",
                "#374FB8",
                "#BD6F26",
                "#4C636F",
              ],
              "hovermode": "closest",
              "legend": Object {
                "orientation": "h",
                "traceorder": "normal",
              },
              "margin": Object {
                "b": 30,
                "l": 30,
                "pad": 4,
                "r": 5,
                "t": 5,
              },
              "showlegend": false,
              "xaxis": Object {
                "automargin": true,
                "fixedrange": false,
                "showgrid": false,
              },
              "yaxis": Object {
                "automargin": true,
                "fixedrange": false,
                "showgrid": false,
              },
            }
          }
          style={
            Object {
              "height": "100%",
              "width": "100%",
            }
          }
          useResizeHandler={true}
        >
          <div
            style={
              Object {
                "height": "100%",
                "width": "100%",
              }
            }
          />
        </PlotlyComponent>
      </Plt>
    </Bar>
  </HorizontalBar>
</div>
`;

exports[`Utils helper functions renders displayVisualization function 4`] = `
<div>
  <HorizontalBar
    layoutConfig={
      Object {
        "xaxis": Object {
          "fixedrange": true,
          "showgrid": false,
        },
        "yaxis": Object {
          "fixedrange": true,
          "showgrid": false,
        },
      }
    }
    visualizations={
      Object {
        "data": Object {
          "count('ip')": Array [],
          "span(timestamp,1h)": Array [],
        },
        "jsonData": Array [],
        "metadata": Object {
          "fields": Array [
            Object {
              "name": "count('ip')",
              "type": "integer",
            },
            Object {
              "name": "span(timestamp,1h)",
              "type": "timestamp",
            },
          ],
        },
        "size": 0,
        "status": 200,
      }
    }
  >
    <Bar
      barConfig={
        Object {
          "orientation": "h",
        }
      }
      layoutConfig={
        Object {
          "xaxis": Object {
            "fixedrange": true,
            "showgrid": false,
          },
          "yaxis": Object {
            "automargin": true,
            "fixedrange": true,
            "showgrid": false,
          },
        }
      }
      visualizations={
        Object {
          "data": Object {
            "count('ip')": Array [],
            "span(timestamp,1h)": Array [],
          },
          "jsonData": Array [],
          "metadata": Object {
            "fields": Array [
              Object {
                "name": "count('ip')",
                "type": "integer",
              },
              Object {
                "name": "span(timestamp,1h)",
                "type": "timestamp",
              },
            ],
          },
          "size": 0,
          "status": 200,
        }
      }
    >
      <Plt
        config={
          Object {
            "orientation": "h",
          }
        }
        data={
          Array [
            Object {
              "marker": Object {
                "color": Array [],
              },
              "name": "count('ip')",
              "orientation": "h",
              "type": "bar",
              "x": Array [],
              "y": Array [],
            },
          ]
        }
        layout={
          Object {
            "colorway": Array [
              "#3CA1C7",
              "#8C55A3",
              "#DB748A",
              "#F2BE4B",
              "#68CCC2",
              "#2A7866",
              "#843769",
              "#374FB8",
              "#BD6F26",
              "#4C636F",
            ],
            "xaxis": Object {
              "automargin": true,
              "fixedrange": true,
              "showgrid": false,
            },
            "yaxis": Object {
              "automargin": true,
              "fixedrange": true,
              "showgrid": false,
            },
          }
        }
      >
        <PlotlyComponent
          config={
            Object {
              "displayModeBar": false,
              "orientation": "h",
            }
          }
          data={
            Array [
              Object {
                "marker": Object {
                  "color": Array [],
                },
                "name": "count('ip')",
                "orientation": "h",
                "type": "bar",
                "x": Array [],
                "y": Array [],
              },
            ]
          }
          debug={false}
          layout={
            Object {
              "autosize": true,
              "barmode": "stack",
              "colorway": Array [
                "#3CA1C7",
                "#8C55A3",
                "#DB748A",
                "#F2BE4B",
                "#68CCC2",
                "#2A7866",
                "#843769",
                "#374FB8",
                "#BD6F26",
                "#4C636F",
              ],
              "hovermode": "closest",
              "legend": Object {
                "orientation": "h",
                "traceorder": "normal",
              },
              "margin": Object {
                "b": 30,
                "l": 30,
                "pad": 4,
                "r": 5,
                "t": 5,
              },
              "showlegend": false,
              "xaxis": Object {
                "automargin": true,
                "fixedrange": true,
                "showgrid": false,
              },
              "yaxis": Object {
                "automargin": true,
                "fixedrange": true,
                "showgrid": false,
              },
            }
          }
          style={
            Object {
              "height": "100%",
              "width": "100%",
            }
          }
          useResizeHandler={true}
        >
          <div
            style={
              Object {
                "height": "100%",
                "width": "100%",
              }
            }
          />
        </PlotlyComponent>
      </Plt>
    </Bar>
  </HorizontalBar>
</div>
`;
